vector/src/main/java/im/vector/app/features/settings/account/deactivation/DeactivateAccountViewModel.kt

```diff
@@ -67,19 +71,22 @@ class DeactivateAccountViewModel @AssistedInject constructor(@Assisted private v
 
         _viewEvents.post(DeactivateAccountViewEvents.Loading())
 
-        session.deactivateAccount(action.password, action.eraseAllData, object : MatrixCallback<Unit> {
-            override fun onSuccess(data: Unit) {
-                _viewEvents.post(DeactivateAccountViewEvents.Done)
-            }
+        viewModelScope.launch {
+            val event = try {
+                session.deactivateAccount(action.password, action.eraseAllData)
+                DeactivateAccountViewEvents.Done
+            } catch (failure: Exception) {
+                if (failure is CancellationException) throw failure
```

Comment:
Why `CancellationException` is treated apart, and why throwing? Will it make the app crash?

Reply:
`CancellationException` in general won't make the app crash.
When `launch` gets a `CancellationException`, it just gracefully stops executing.
When `async` gets a `CancellationException`, it just gracefully stops executing and other coroutines calling `await()` will get the cancellation exception.

The only case when it may crash the app is if it is **directly** thrown inside `runBlocking`, since it has to run in non-coroutine code.

Reply:
Ok, thanks for the answer!

Reply:
Isn't it ok to let the CancellationException go to OtherFailure then? it won't change anything in the execution flow, but we'll loose the error in the UI if not posting to viewEvents

Reply:
It's technically not an error, the task has just been cancelled and this will only happen if the view model has been cleared, so if the UI has been destroyed.
But yeah it doesn't make much of a difference I can remove the special casing.

Reply:
Yes I wasn't thinking about this specific case, but more generally. We can get a cancellation if some suspend got cancelled by any inner stuff. By the way, we still have to be able to cancel all session coroutine (clearing cache/logout)

Reply:
That complicates things..... Is that the only time when you cancel all sessions coroutines?
In the case of `viewModelScope` the coroutine would already have been cancelled before user logs out or clears the cache, no?

Reply:
I guess for most of the cases it will be ok yeah